students/ScottL/session05/Mailroom_Part3.py

```diff
@@ -0,0 +1,141 @@
+        try:
+            total_donation = round(sum(global_donation_list[item]))
+        except TypeError:
+            print("Error: A value other than a number is listed as a "
+                  "donation 'amount' for '{}'.".format(item))
+        else:
+            number_of_donations = len(global_donation_list[item])
+            try:
+                sorted_donation_list.append([item, total_donation, number_of_donations,
+                                            round(total_donation / number_of_donations)])
+            except ZeroDivisionError:
+                sorted_donation_list.append([item, total_donation, number_of_donations, 0])
+    sorted_donation_list.sort(key=lambda x: x[1], reverse=True)
+    return sorted_donation_list
+
+
+def display_donor_names():
+    """Print the list of full donor names for the user."""
+    print("\n", DONOR_LIST_HEADER)
+    for entry in global_donation_list:
+        print(entry)
+
+
+def draft_email(name, amount):
+    """
+    Display text reminiscent of an email thanking a donor.
+    :param name: donor name
+    :param amount: amount of donation
+    """
+    print("\nDear {},\nThanks for the generous donation of ${:,.2f}."
+          "\n\nSincerely,\nFundraising Team".format(name, amount))
+
+
+def input_donation_amount(donor_name):
+    """
+    Receive input from user on amount of donation. Process accordingly.
+    :param donor_name: name of donor
+    """
+    donation_amount = input("Please enter the donation amount "
+                            "(or enter 'Quit' to return to Main Menu): ").strip("$")
+    if donation_amount.strip().lower() != "quit":
+        try:
+            donation_amount = float(donation_amount)
+        except ValueError:
+            print("The entry for 'amount' count was not a valid number. Please try again.")
+            input_donation_amount(donation_amount)
+        else:
+            if donor_name in global_donation_list:
+                global_donation_list[donor_name].append(float(donation_amount))
+            else:
+                global_donation_list[donor_name] = [float(donation_amount)]
+            draft_email(donor_name.title(), float(donation_amount))
+
+
+def input_donor_name():
+    """Prompt the user to enter the donor name."""
+    donor_name = input("\nPlease enter the full name of the donor, "
+                       "'List' to see a list of existing donors, "
+                       "or 'Quit' to return to the Main Menu: ")
+    if donor_name.strip().lower() != "quit":
+        if donor_name.strip().lower() == "list":
+            display_donor_names()
+            input_donor_name()
+        else:
+            input_donation_amount(donor_name)
+
+
+def display_report():
+    """Call the function to calculate total & avg donations and then display a report."""
+    print(REPORT_HEADER_1, REPORT_HEADER_2)
+    for item in donation_calcs():
+        print("{:<26}${:>12,}{:^17}${:>11,}".format(item[0], item[1], item[2], item[3]))
```

Comment:
Here is a good example of where not to use it as the print line is very long and it wouldn't make the comp very readable.